### PR TITLE
http: use Header struct for headers

### PR DESCRIPTION
**Pull request by mcastorina**
This PR replaces `headers map[string]string` and `lheaders map[string]string` in http with the already implemented `Header` struct.

### CODE REVIEW

File: vlib/net/http/http.v
Comment (mcastorina) [inline vlib/net/http/http.v:195]: Thoughts on changing this signature to use an enum instead of a string? I thought it would be better to encourage using the enum keys instead of strings, but I would like others' opinions.

Comment (AliChraghi) [reply vlib/net/http/http.v:195]: i think use string is better, bc use lower memory and more flexible

Comment (JalonSolov) [reply vlib/net/http/http.v:195]: More consistent if you use enums, as well as not having to worry about letter case in strings.

**Review by AliChraghi — CHANGES_REQUESTED** [review]
be sure use string instead enum

File: vlib/net/http/http.v
Comment (mcastorina) [reply vlib/net/http/http.v:195]: @AliChraghi using strings would be *more* memory and slower because not any string can be a header key, so it has to be checked. These checks are skipped with the enum because we know it's already a valid header key.

It's still possible to add a custom header with:
```v
req.headers.add_str('Custom-Header', 'value')
```

Maybe I should add another method to Request.. `add_header_str` ?

Comment (AliChraghi) [reply vlib/net/http/http.v:195]: true, I do not see another problem

**Review by AliChraghi — APPROVED** [review]

**Comment by mcastorina** [comment]
PR for gitly once this is merged: https://github.com/vlang/gitly/pull/129

File: vlib/net/http/http.v
Comment (medvednikov) [inline vlib/net/http/http.v:22]: plural vs singular inconsistency

in Go it's `header Header`

Comment (mcastorina) [reply vlib/net/http/http.v]: Updated to `header Header`

**Comment by medvednikov** [comment]
What if a custom header is needed?

**Comment by mcastorina** [comment]
@medvednikov a custom string header can be added by accessing the `header` attribute directly: 

```v
req.header.add_str('key', 'value') or { panic(err) }
```

If that's too inconsistent, I can add another method on `Request`

```v
pub fn (mut req Request) add_header_str(key string, value string) ?
```

**Comment by JalonSolov** [comment]
How about `add_custom_header`, since that's what it would be used for.

**Comment by spytheman** [comment]
I agree, `add_custom_header` is more meaningful than `add_header_str`.

**Comment by mcastorina** [comment]
I've added `add_custom_header` but that raises the question of whether the `add_str` method of `Header` should be changed too?

**Comment by JalonSolov** [comment]
Existing `add` looks like what `add_header` should be.  `add_str` looks like a way to add a header as a string... which is basically what `add_custom_header` would be, _except_ for the `canonicalize` call.

I understand wanting to keep the headers "pretty", but... headers are case-insensitive according to the HTTP standard, so I don't see a need to force them to be any particular case or pattern.  At most, I would do the check to see if lowercase of what was sent in is a standard header, and use it, just for consistency on those, but any custom headers should be left alone.  In other words, most of the `canonicalize` routine could go away, and likely be renamed as well to match just what is left.

We have no idea of someone actually does a case-SENSITIVE check on "MY-CUSTOM-header".

All that said, the routines could just the `add` (using the existing method) and `add_custom`, since they are receiver methods on Header already.

**Comment by mcastorina** [comment]
I see your point, but I made `Header` that way for a few reasons:

1. That's the way Go's `Header` behaves (with canonicalization). I know V isn't Go and because they do it one way doesn't mean we should too, but that was the model I chose.
2. It standardizes the format of the headers we send out, so if you do `header.add_str('key', 'a')` and `header.add_str('KEY', 'a')` it will look the same. I agree the specific format doesn't matter (we could lowercase it)
3. By canonicalizing before we add it to the map we eliminate the need to keep two maps (one for lowercase, one for exactly what the user used).

> headers are case-insensitive according to the HTTP standard

> We have no idea if someone actually does a case-SENSITIVE check on "MY-CUSTOM-header"

Whoever is doing case sensitive checks is not following the HTTP standard then.

**Comment by dumblob** [comment]
>Whoever is doing case sensitive checks is not following the HTTP standard then.

This is more than true. **However** *V* has it's goal to be useful **by default** (i.e. without writing your own HTTP library) even for low level programming and e.g. to circumvent most deep packet inspection (DPI) systems it's enough to make the header case somehow "unexpected". In other words, there are legitimate use cases when this "non-standard" but still "correct" behavior makes sense.

Speaking of uniformity (which inevitably affects performance), have you @mcastorina considered designing a second smaller but full featured HTTP API different than the Go-like API with full zero-copy semantics (i.e. deferred sax-like parsing and copy-on-write behavior)? Would it make sense to have it along the current Go-like API for truly high-performance use cases?

**Comment by mcastorina** [comment]
That's a good example use-case, thank you for sharing @dumblob. What would be the expected output for something like this..
```v
mut header := http.new_header()
header.add(.accept, 'text/html')
header.add_str('accept', 'image/*') ?
```

I think the average user would expect `Accept: text/html; image/*` as opposed to
```
Accept: text/html
accept: image/*
```
but maybe that's just me and the latter should be the behavior?

---

> have you @mcastorina considered designing a second smaller but full featured HTTP API

> Would it make sense to have it along the current Go-like API for truly high-performance use cases?

I have not, but I think that would make for a good module. Having both in vlib goes against V's "one way"

**Comment by JalonSolov** [comment]
I think he was just agreeing that any custom headers which DON'T match one of the standard one should be left in whatever case the user gives, instead of being coerced into some arbitrary "canonicalized" version.  If the same one is given multiple times, the case should stay as whatever the first one was, but act additively as your example:
```v
mut header := http.new_header()
header.add(.accept, 'text/html')
header.add_custom('accept', 'image/*') ?
```
Result would be `Accept: text/html; image/*` since `Accept` is a standard header.
```v
mut header := http.new_header()
header.add_custom('XYZZY', 'text/html')
header.add_custom('xyzzy', 'image/*') ?
```
would result in `XYZZY: text/html; image/*` since the all uppercase version was specified first.

**Comment by mcastorina** [comment]
Okay, I can add those changes to this PR. What are thoughts on custom headers that *do* match a standard one? For consistency I would say it should keep what the user specified first.

```v
header.add_custom('accept', 'text/html') ?
header.add(.accept, 'image/*')
```
```
accept: text/html; image/*
```

**Comment by JalonSolov** [comment]
On that one, I'd go with the standard.  Otherwise what's the use of having a standard, or enums, etc.

**Comment by dumblob** [comment]
> ```v
> mut header := http.new_header()
> header.add(.accept, 'text/html')
> header.add_str('accept', 'image/*') ?
> ```
> 
> I think the average user would expect `Accept: text/html; image/*` as opposed to
> 
> ```
> Accept: text/html
> accept: image/*
> ```
> 
> but maybe that's just me and the latter should be the behavior?

I'd definitely prefer the latter. Simply because if I wanted it to be "squashed", then I'd use `header.add(.accept, 'image/*')`. In all other cases, the code should do what it explicitly says - i.e. take headers **verbatim**. So **no squashing** even if the order was reversed (i.e. `header.add_str('accept', 'image/*')` first and then `header.add(.accept, 'text/html')`.

If squashing is desired for headers not supported by the `http` *V* module, then either custom enum with `.myheader` shall be added and `header.add()` shall handle it appropriately (maybe generics will be needed, maybe other tricks, ... - I'll let this to others to decide) or simply no such support will be provided and the programmer will be expected to do it on her own with what *V* language offers :wink:.

**Comment by JalonSolov** [comment]
Since headers are supposed to be treated case-insensitively, how would you expect
```
Accept: text/html
accept: image/*
```
to be useful?  I would be willing to bet the second (case-insensitive, remember!) value to overwrite the first, not add to it.  So basically it would be useless to have the first one there.

**Comment by mcastorina** [comment]
I can go either way when it comes to creating a HTTP request, however this struct is also used for handling *incoming* HTTP requests.

If a client sends `accept: text/html` or `ACCEPT: text/html`, I want to just do `header.get(.accept)` and that's it.

**Comment by dumblob** [comment]
>If a client sends `accept: text/html` or `ACCEPT: text/html`, I want to just do `header.get(.accept)` and that's it.

That doesn't sound good enough to me - IMHO it violates the semantics HTTP has. HTTP headers are a "multimap" - i.e. the same key can be there any number of times. And I think it's important to expose exactly this semantics through the API.

Maybe `header.get_first( .accept )` (and `header.get_first( 'accept' )`) would return exactly one string value, but `header.get( 'Accept' )` (`header.get( .accept )` wouldn't be allowed - instead a `header_to_str(.accept)` would need to be used) would return an array of all `Accept` (not `ACCEPT` nor `accept` nor anything else) values and finally `header.getsimilar( 'accept' )` would return an array of everything satisfying equality comparison after performing `tolower()` on both LHS and RHS.

**Comment by JalonSolov** [comment]
From HTTP 1.1 spec, section 4.2:
```
Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.
```

**Comment by dumblob** [comment]
@JalonSolov sure, that's a clear definition how to **interpret** already existing (usually received) data, but it does **not** dictate how to construct them nor how to send them.

All these HTTP standards are written around the first part of the good old advice:

>Tolerate as much as you can on input, produce as strict as possible on output.

**Comment by JalonSolov** [comment]
Likely should set up a general vote.  I'd prefer what I said... for standard headers, coerce them to the same case, and combine the values.  For all others, leave them as the user typed them in, but still combine values if they give multiples of the same thing (keeping the initial casing).

@dumblob appears to think it should be "open season" to put whatever casing you want, and keep them separate, whether it's a standard header or not.  In that case, there's no real reason to have the standard header enum, either.  Just take whatever's given and add it to the headers.

**Comment by dumblob** [comment]
Sure, we can make a vote. I don't think though we have aligned goals. My goal is to have the **standard library** "universal" (not performant as we clarified above). If a standard library is not universal, then people will be forced to again and again write their own and that's never good a language (especially considering the *V* mantra to do one thing one way - but here too many would use their own homegrown libs).

So, I'd still prefer either (1) or (2) from @mcastorina 's proposal.

Some last thoughts about the topic. If RFC guys wanted everybody to **create** coerced values with verbatim headers, they'd specify so (but they didn't and we have the choice of coercion as well as case insensitivity). Also the behavior @JalonSolov proposes is a complex opaque (non-explicit) behavior - as a newcomer I'd be seriously surprised by the result of what I've written (and I'd probably first find out after some accident because I wouldn't come up with the idea to unit test calls to **standard language library**).

**Comment by mcastorina** [comment]
Alright, let's vote then.

:heart: Coerce common headers to the same case; leave custom headers in the original case
:rocket: Coerce all headers to the same case
:tada: No coersion; provide methods for exact and "fuzzy" retrieval

@JalonSolov please let me know if there is a better venue for voting

**Pull request merged**
